neqo-transport/src/connection/tests/datagram.rs

```diff
@@ -0,0 +1,243 @@
+    assert_eq!(server.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let out = server.process_output(now()).dgram().unwrap();
+    assert_eq!(server.stats().frame_tx.datagram, dgram_sent + 1);
+
+    client.process_input(out, now());
+    let datagram =
+        |e| matches!(e, ConnectionEvent::Datagram(data) if data == DATA_SMALLER_THAN_MTU);
+    assert!(client.events().any(datagram));
+}
+
+#[test]
+fn datagram_enabled_on_server() {
+    let mut client = default_client();
+    let mut server = new_server(
+        ConnectionParameters::default().max_datagram_frame_size(DATAGRAM_LEN_SMALLER_THAN_MTU),
+    );
+    connect_force_idle(&mut client, &mut server);
+
+    assert_eq!(client.max_dgram_size(), Ok(DATAGRAM_LEN_SMALLER_THAN_MTU));
+    assert_eq!(server.max_dgram_size(), Err(Error::NotAvailable));
+    assert_eq!(
+        server.send_dgram(&DATA_SMALLER_THAN_MTU),
+        Err(Error::NotAvailable)
+    );
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let out = client.process_output(now()).dgram().unwrap();
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    server.process_input(out, now());
+    let datagram =
+        |e| matches!(e, ConnectionEvent::Datagram(data) if data == DATA_SMALLER_THAN_MTU);
+    assert!(server.events().any(datagram));
+}
+
+fn create_and_connect_with_dgrams() -> (Connection, Connection) {
+    let mut client =
+        new_client(ConnectionParameters::default().max_datagram_frame_size(DATAGRAM_LEN_MAX));
+    let mut server =
+        new_server(ConnectionParameters::default().max_datagram_frame_size(DATAGRAM_LEN_MAX));
+    connect_force_idle(&mut client, &mut server);
+    (client, server)
+}
+
+#[test]
+fn mtu_limit() {
+    let (client, server) = create_and_connect_with_dgrams();
+
+    assert_eq!(client.max_dgram_size(), Ok(DATAGRAM_LEN_MTU));
+    assert_eq!(server.max_dgram_size(), Ok(DATAGRAM_LEN_MTU));
+}
+
+#[test]
+fn limit_data_size() {
+    let (mut client, mut server) = create_and_connect_with_dgrams();
+
+    assert!(u64::try_from(DATA_BIGGER_THAN_MTU.len()).unwrap() > DATAGRAM_LEN_MTU);
+    assert_eq!(
+        client.send_dgram(DATA_BIGGER_THAN_MTU),
+        Err(Error::TooMuchData)
+    );
+    assert_eq!(
+        server.send_dgram(DATA_BIGGER_THAN_MTU),
+        Err(Error::TooMuchData)
+    );
+}
+
+#[test]
+fn datagram_acked() {
+    let (mut client, mut server) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let out = client.process_output(now()).dgram();
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    let dgram_received = server.stats().frame_rx.datagram;
+    server.process_input(out.unwrap(), now());
+    assert_eq!(server.stats().frame_rx.datagram, dgram_received + 1);
+    let now = now() + Duration::from_millis(1);
```

Comment:
Don't put fixed timeouts in.  Is this the ack delay?

I think that this test might be fragile.  It relies on the server needing to send an ACK for other reasons.

Reply:
The server needs to send an ack for the received datagram. I will use AT_LEAST_PTO time here.